### PR TITLE
Add basic ETL job for search dashboard dataset

### PR DESCRIPTION
**Pull request by harterrt**
[Example notebook](https://github.com/mozilla/search-adhoc-analysis/blob/master/search-dashboard/search_dataset_test.ipynb)

### CODE REVIEW

File: tests/test_search_dashboard.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    return define_dataframe_factory(map(to_field, [
+        ('submission_date', '20170101',   StringType(), False),
+        ('country',         'DE',         StringType(), True),
+        ('app_version',     '54.0.1',     StringType(), True),
+        ('distribution_id', None,         StringType(), True),
+        ('engine',          'google',     StringType(), False),
+        ('source',          'urlbar',     StringType(), False),
+        ('search_count',    4,            LongType(),   False),
+        ('type',            'chrome-sap', StringType(), False),
+    ]))([
```
Comment (acmiyaguchi) [inline tests/test_search_dashboard.py]: Nit: Split this into two statements to be easier to read.

File: mozetl/search/dashboard.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        .withColumn(exploded_col_name, explode(col('search_counts')))
+        .withColumn(*define_sc_column('engine'))
+        .withColumn(*define_sc_column('source'))
+        .withColumn(*define_sc_column('count'))
+        .drop(exploded_col_name)
+        .drop('search_counts')
+    )
+
+
+def add_derived_columns(exploded_search_counts):
```
Comment (acmiyaguchi) [inline mozetl/search/dashboard.py]: Add a more descriptive function name like `with_derived_search_type` or a comment related to the type of derived column that it adds to the dataframe.

Comment (harterrt) [inline mozetl/search/dashboard.py]: I plan on adding about a dozen new columns to this function in the next revision. Accordingly, I opted for a comment.

File: tests/conftest.py [hunk 3]
```diff
@@ -27,6 +27,17 @@ def nop(*args, **kwargs):
     monkeypatch.setattr("pyspark.sql.SparkSession.stop", nop)
 
 
+@pytest.fixture
+def df_equals(row_to_dict):
+    def to_comparable(df):
+        return sorted(map(row_to_dict, df.collect()))
```
Comment (harterrt) [inline tests/conftest.py]: `sorted` is being called on a `list` of `dicts` where each element of the list represents one row of a dataframe.

File: tests/test_search_dashboard.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    return {
+        'engine': engine,
+        'source': source,
+        'count':  count,
+    }
+
+
+@pytest.fixture()
+def generate_main_summary_data(define_dataframe_factory):
+    return define_dataframe_factory(map(to_field, [
```
Comment (harterrt) [inline tests/test_search_dashboard.py]: I am torn. I had it inside `define_dataframe_factory` but decide to remove it. The `dataframe_field` class serves two purposes: 1) allowing easy access to each field and 2) documenting what each field means.

I was concerned the nameless tuples would be difficult to understand without keeping `to_field` close to the tuple definition. If you don't think this helps readability, I'm happy to move this back into `define_dataframe_factory`

File: tests/test_search_dashboard.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        ('submission_date', '20170101', StringType(), False),
```
Comment (harterrt) [inline tests/test_search_dashboard.py]: If it's all the same to you, I'd prefer to keep StringType() for consistency with the other types.

File: tests/test_search_dashboard.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        ('country',         'DE',       StringType(), True),
+        ('app_version',     '54.0.1',   StringType(), True),
+        ('distribution_id', None,       StringType(), True),
+        ('ignored_col',     1.0,        DoubleType(), True),
+        (
+            'search_counts',
+            [generate_search_count()],
+            ArrayType(StructType([
```
Comment (harterrt) [inline tests/test_search_dashboard.py]: Indeed. Much nicer.

File: tests/test_search_dashboard.py (hunk 1 above)
Comment (harterrt) [inline tests/test_search_dashboard.py]: Done.

File: mozetl/search/dashboard.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            when(col('source').startswith('sap:'), 'in-content-sap')
+            .otherwise(
+                when(col('source').startswith('follow-on:'), 'follow-on')
+                .otherwise('chrome-sap')
+            )
+        )
+    )
+
+
+
```
Comment (harterrt) [inline mozetl/search/dashboard.py]: Done

File: mozetl/search/dashboard.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    return (
+        main_summary
+        .select(input_columns)
+        .withColumn(exploded_col_name, explode(col('search_counts')))
+        .withColumn(*define_sc_column('engine'))
+        .withColumn(*define_sc_column('source'))
+        .withColumn(*define_sc_column('count'))
+        .drop(exploded_col_name)
+        .drop('search_counts')
```
Comment (harterrt) [inline mozetl/search/dashboard.py]: Done.

File: mozetl/search/dashboard.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
+import click
```
Comment (harterrt) [inline mozetl/search/dashboard.py]: Done.

**Comment by harterrt** [comment]
Thanks, @acmiyaguchi. Ready for another look.

**Comment by acmiyaguchi** [comment]
Looks good to me, all that needs to be done is to fix the small linting error:

```
mozetl/search/dashboard.py:76:1: E303 too many blank lines (3)
```

**Comment by harterrt** [comment]
Doh. Fixed. Merging once tests pass.

**Pull request merged**